client/dashboard/domains/domain-contact-details/region-address-fieldsets.tsx

```diff
@@ -0,0 +1,194 @@
+import { type StatesListItem, type DomainContactDetails } from '@automattic/api-core';
+import { __experimentalInputControl as InputControl, SelectControl } from '@wordpress/components';
+import { useViewportMatch } from '@wordpress/compose';
+import { type Field, type DataFormControlProps } from '@wordpress/dataviews';
+import { __ } from '@wordpress/i18n';
+import { useEffect } from 'react';
+import {
+	CHECKOUT_EU_ADDRESS_FORMAT_COUNTRY_CODES,
+	CHECKOUT_UK_ADDRESS_FORMAT_COUNTRY_CODES,
+	type CountryListItem,
+} from './custom-form-fieldsets/constants';
+import { getCountryPostalCodeSupport } from './custom-form-fieldsets/get-country-postal-code-support';
+
+const getPostalCodeLabel = ( countryCode: string ): string => {
+	switch ( countryCode ) {
+		case 'US':
+			return __( 'ZIP code' );
+		default:
+			return __( 'Postal Code' );
+	}
+};
+
+const getStateSelectLabel = ( countryCode: string ): string => {
+	switch ( countryCode ) {
+		case 'CA':
+			return __( 'Select Province' );
+		case 'US':
+			return __( 'Select State' );
+		default:
+			return __( 'Select State' );
+	}
+};
+
+const createStateFieldEdit = ( statesList: StatesListItem[] | undefined, countryCode: string ) => {
+	const StateFieldEdit = ( {
+		field,
+		onChange,
+		data,
+		hideLabelFromVision,
+	}: DataFormControlProps< DomainContactDetails > ) => {
+		const { id, getValue } = field;
+		const currentValue = getValue?.( { item: data } );
+
+		// If the item data is not in the statesList, set the state to the first option
+		useEffect( () => {
+			if ( statesList && statesList.length > 0 ) {
+				if ( ! statesList.some( ( state ) => state.code === currentValue ) ) {
+					onChange( { [ id ]: statesList[ 0 ]?.code } );
+				}
+			}
+		}, [ currentValue, onChange, id ] );
+
+		const stateLabel = getStateSelectLabel( countryCode );
+
+		if ( ! statesList || statesList.length === 0 ) {
+			return (
+				<InputControl
+					__next40pxDefaultSize
+					label={ hideLabelFromVision ? '' : stateLabel }
+					placeholder={ __( 'State' ) }
```

Comment:
Do we need the placeholder? I don't see the other fields having one. Maybe remove it for consistency?